lib/libmenu/menu.3
'\" t
.\" $OpenBSD: menu.3,v 1.4 2019/02/13 07:18:58 nicm Exp $
.\"
.\"***************************************************************************
.\" Copyright 2018-2021,2023 Thomas E. Dickey                                *
.\" Copyright 1998-2014,2017 Free Software Foundation, Inc.                  *
.\"                                                                          *
.\" Permission is hereby granted, free of charge, to any person obtaining a  *
.\" copy of this software and associated documentation files (the            *
.\" "Software"), to deal in the Software without restriction, including      *
.\" without limitation the rights to use, copy, modify, merge, publish,      *
.\" distribute, distribute with modifications, sublicense, and/or sell       *
.\" copies of the Software, and to permit persons to whom the Software is    *
.\" furnished to do so, subject to the following conditions:                 *
.\"                                                                          *
.\" The above copyright notice and this permission notice shall be included  *
.\" in all copies or substantial portions of the Software.                   *
.\"                                                                          *
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
.\" IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
.\" DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
.\" OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
.\" THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
.\"                                                                          *
.\" Except as contained in this notice, the name(s) of the above copyright   *
.\" holders shall not be used in advertising or otherwise to promote the     *
.\" sale, use or other dealings in this Software without prior written       *
.\" authorization.                                                           *
.\"***************************************************************************
.\"
.\" $Id: menu.3,v 1.5 2023/10/17 09:52:10 nicm Exp $
.TH menu 3 2023-08-19 "ncurses 6.4" "Library calls"
.ie \n(.g .ds `` \(lq
.el       .ds `` ``
.ie \n(.g .ds '' \(rq
.el       .ds '' ''
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.SH NAME
\fBmenu\fP \- curses extension for programming menus
.SH SYNOPSIS
\fB#include <menu.h>\fP
.SH DESCRIPTION
The \fBmenu\fP library provides terminal-independent facilities for composing
menu systems on character-cell terminals.
The library includes: item routines,
which create and modify menu items; and menu routines, which group items into
menus, display menus on the screen, and handle interaction with the user.
.PP
The \fBmenu\fP library uses the \fBcurses\fP libraries, and a curses
initialization routine such as \fBinitscr\fP must be called before using any of
these functions.
To use the \fBmenu\fP library, link with the options
\fB\-lmenu \-lcurses\fP.
.
.SS Current Default Values for Item Attributes
.
The \fBmenu\fP library maintains a default value for item attributes.
You can
get or set this default by calling the appropriate \fBget_\fP or \fBset_\fP
routine with a \fBNULL\fP item pointer.
Changing this default with a
\fBset_\fP function affects future item creations, but does not change the
rendering of items already created.
.
.SS Routine Name Index
.
The following table lists each \fBmenu\fP routine and the name of
the manual page on which it is described.
.PP
.TS
l l .
\fBcurses\fP Routine Name	Manual Page Name
=
current_item	\fBmitem_current\fP(3)
free_item 	\fBmitem_new\fP(3)
free_menu 	\fBmenu_new\fP(3)
item_count	\fBmenu_items\fP(3)
item_description	\fBmitem_name\fP(3)
item_index	\fBmitem_current\fP(3)
item_init 	\fBmenu_hook\fP(3)
item_name 	\fBmitem_name\fP(3)
item_opts 	\fBmitem_opts\fP(3)
item_opts_off	\fBmitem_opts\fP(3)
item_opts_on	\fBmitem_opts\fP(3)
item_term 	\fBmenu_hook\fP(3)
item_userptr	\fBmitem_userptr\fP(3)
item_value	\fBmitem_value\fP(3)
item_visible	\fBmitem_visible\fP(3)
menu_back 	\fBmenu_attributes\fP(3)
menu_driver	\fBmenu_driver\fP(3)
menu_fore 	\fBmenu_attributes\fP(3)
menu_format	\fBmenu_format\fP(3)
menu_grey 	\fBmenu_attributes\fP(3)
menu_init 	\fBmenu_hook\fP(3)
menu_items	\fBmenu_items\fP(3)
menu_mark 	\fBmenu_mark\fP(3)
menu_opts 	\fBmenu_opts\fP(3)
menu_opts_off	\fBmenu_opts\fP(3)
menu_opts_on	\fBmenu_opts\fP(3)
menu_pad  	\fBmenu_attributes\fP(3)
menu_pattern	\fBmenu_pattern\fP(3)
menu_request_by_name	\fBmenu_requestname\fP(3)
menu_request_name	\fBmenu_requestname\fP(3)
menu_spacing	\fBmenu_spacing\fP(3)
menu_sub  	\fBmenu_win\fP(3)
menu_term 	\fBmenu_hook\fP(3)
menu_userptr	\fBmenu_userptr\fP(3)
menu_win  	\fBmenu_win\fP(3)
new_item  	\fBmitem_new\fP(3)
new_menu  	\fBmenu_new\fP(3)
pos_menu_cursor	\fBmenu_cursor\fP(3)
post_menu 	\fBmenu_post\fP(3)
scale_menu	\fBmenu_win\fP(3)
set_current_item	\fBmitem_current\fP(3)
set_item_init	\fBmenu_hook\fP(3)
set_item_opts	\fBmitem_opts\fP(3)
set_item_term	\fBmenu_hook\fP(3)
set_item_userptr	\fBmitem_userptr\fP(3)
set_item_value	\fBmitem_value\fP(3)
set_menu_back	\fBmenu_attributes\fP(3)
set_menu_fore	\fBmenu_attributes\fP(3)
set_menu_format	\fBmenu_format\fP(3)
set_menu_grey	\fBmenu_attributes\fP(3)
set_menu_init	\fBmenu_hook\fP(3)
set_menu_items	\fBmenu_items\fP(3)
set_menu_mark	\fBmenu_mark\fP(3)
set_menu_opts	\fBmitem_opts\fP(3)
set_menu_pad	\fBmenu_attributes\fP(3)
set_menu_pattern	\fBmenu_pattern\fP(3)
set_menu_spacing	\fBmenu_spacing\fP(3)
set_menu_sub	\fBmenu_win\fP(3)
set_menu_term	\fBmenu_hook\fP(3)
set_menu_userptr	\fBmenu_userptr\fP(3)
set_menu_win	\fBmenu_win\fP(3)
set_top_row	\fBmitem_current\fP(3)
top_row   	\fBmitem_current\fP(3)
unpost_menu	\fBmenu_post\fP(3)
.TE
.SH RETURN VALUE
Routines that return pointers return \fBNULL\fP on error.
Routines that return
an integer return one of the following error codes:
.TP 5
.B E_OK
The routine succeeded.
.TP 5
.B E_BAD_ARGUMENT
Routine detected an incorrect or out-of-range argument.
.TP 5
.B E_BAD_STATE
Routine was called from an initialization or termination function.
.TP 5
.B E_NO_MATCH
Character failed to match.
.TP 5
.B E_NO_ROOM
Menu is too large for its window.
.TP 5
.B E_NOT_CONNECTED
No items are connected to the menu.
.TP 5
.B E_NOT_POSTED
The menu has not been posted.
.TP 5
.B E_NOT_SELECTABLE
The designated item cannot be selected.
.TP 5
.B E_POSTED
The menu is already posted.
.TP 5
.B E_REQUEST_DENIED
The menu driver could not process the request.
.TP 5
.B E_SYSTEM_ERROR
System error occurred (see \fBerrno\fP(3)).
.TP 5
.B E_UNKNOWN_COMMAND
The menu driver code saw an unknown request code.
.SH NOTES
The header file \fB<menu.h>\fP automatically includes the header files
\fB<curses.h>\fP and \fB<eti.h>\fP.
.PP
In your library list, libmenu.a should be before libncurses.a; that is,
you should say \*(``\-lmenu \-lncurses\*('', not the other way around
(which would give a link-error when using static libraries).
.SH PORTABILITY
These routines emulate the System V menu library.
They were not supported on
Version 7 or BSD versions.
.PP
The menu facility was documented in SVr4.2 in
\fICharacter User Interface Programming (UNIX SVR4.2)\fP.
.PP
It is not part of X/Open Curses.
.PP
Aside from ncurses, there are few implementations:
.bP
systems based on SVr4 source code, e.g., Solaris.
.bP
NetBSD curses.
.SH AUTHORS
Juergen Pfeifer.
Manual pages and adaptation for ncurses by Eric S. Raymond.
.SH SEE ALSO
\fBcurses\fP(3) and related pages whose names begin \*(``menu_\*(''
for detailed descriptions of the entry points.
.PP
This describes \fBncurses\fP
version 6.4 (patch 20230826).


usr.bin/tmux/menu.c
/* $OpenBSD: menu.c,v 1.53 2024/08/21 04:17:09 nicm Exp $ */

/*
 * Copyright (c) 2019 Nicholas Marriott <nicholas.marriott@gmail.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF MIND, USE, DATA OR PROFITS, WHETHER
 * IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>

#include <stdlib.h>
#include <string.h>

#include "tmux.h"

struct menu_data {
	struct cmdq_item	*item;
	int			 flags;

	struct grid_cell	 style;
	struct grid_cell	 border_style;
	struct grid_cell	 selected_style;
	enum box_lines		 border_lines;

	struct cmd_find_state	 fs;
	struct screen		 s;

	u_int			 px;
	u_int			 py;

	struct menu		*menu;
	int			 choice;

	menu_choice_cb		 cb;
	void			*data;
};

void
menu_add_items(struct menu *menu, const struct menu_item *items,
    struct cmdq_item *qitem, struct client *c, struct cmd_find_state *fs)
{
	const struct menu_item	*loop;

	for (loop = items; loop->name != NULL; loop++)
		menu_add_item(menu, loop, qitem, c, fs);
}

void
menu_add_item(struct menu *menu, const struct menu_item *item,
    struct cmdq_item *qitem, struct client *c, struct cmd_find_state *fs)
{
	struct menu_item	*new_item;
	const char		*key = NULL, *cmd, *suffix = "";
	char			*s, *trimmed, *name;
	u_int			 width, max_width;
	int			 line;
	size_t			 keylen, slen;

	line = (item == NULL || item->name == NULL || *item->name == '\0');
	if (line && menu->count == 0)
		return;
	if (line && menu->items[menu->count - 1].name == NULL)
		return;

	menu->items = xreallocarray(menu->items, menu->count + 1,
	    sizeof *menu->items);
	new_item = &menu->items[menu->count++];
	memset(new_item, 0, sizeof *new_item);

	if (line)
		return;

	if (fs != NULL)
		s = format_single_from_state(qitem, item->name, c, fs);
	else
		s = format_single(qitem, item->name, c, NULL, NULL, NULL);
	if (*s == '\0') { /* no item if empty after format expanded */
		menu->count--;
		return;
	}
	max_width = c->tty.sx - 4;

	slen = strlen(s);
	if (*s != '-' && item->key != KEYC_UNKNOWN && item->key != KEYC_NONE) {
		key = key_string_lookup_key(item->key, 0);
		keylen = strlen(key) + 3; /* 3 = space and two brackets */

		/*
		 * Add the key if it is shorter than a quarter of the available
		 * space or there is space for the entire item text and the
		 * key.
		 */
		if (keylen <= max_width / 4)
			max_width -= keylen;
		else if (keylen >= max_width || slen >= max_width - keylen)
			key = NULL;
	}

	if (slen > max_width) {
		max_width--;
		suffix = ">";
	}
	trimmed = format_trim_right(s, max_width);
	if (key != NULL) {
		xasprintf(&name, "%s%s#[default] #[align=right](%s)",
		    trimmed, suffix, key);
	} else
		xasprintf(&name, "%s%s", trimmed, suffix);
	free(trimmed);

	new_item->name = name;
	free(s);

	cmd = item->command;
	if (cmd != NULL) {
		if (fs != NULL)
			s = format_single_from_state(qitem, cmd, c, fs);
		else
			s = format_single(qitem, cmd, c, NULL, NULL, NULL);
	} else
		s = NULL;
	new_item->command = s;
	new_item->key = item->key;

	width = format_width(new_item->name);
	if (*new_item->name == '-')
		width--;
	if (width > menu->width)
		menu->width = width;
}

struct menu *
menu_create(const char *title)
{
	struct menu	*menu;

	menu = xcalloc(1, sizeof *menu);
	menu->title = xstrdup(title);
	menu->width = format_width(title);

	return (menu);
}

void
menu_free(struct menu *menu)
{
	u_int	i;

	for (i = 0; i < menu->count; i++) {
		free((void *)menu->items[i].name);
		free((void *)menu->items[i].command);
	}
	free(menu->items);

	free((void *)menu->title);
	free(menu);
}

struct screen *
menu_mode_cb(__unused struct client *c, void *data, u_int *cx, u_int *cy)
{
	struct menu_data	*md = data;

	*cx = md->px + 2;
	if (md->choice == -1)
		*cy = md->py;
	else
		*cy = md->py + 1 + md->choice;

	return (&md->s);
}

/* Return parts of the input range which are not obstructed by the menu. */
void
menu_check_cb(__unused struct client *c, void *data, u_int px, u_int py,
    u_int nx, struct overlay_ranges *r)
{
	struct menu_data	*md = data;
	struct menu		*menu = md->menu;

	server_client_overlay_range(md->px, md->py, menu->width + 4,
	    menu->count + 2, px, py, nx, r);
}

void
menu_draw_cb(struct client *c, void *data,
    __unused struct screen_redraw_ctx *rctx)
{
	struct menu_data	*md = data;
	struct tty		*tty = &c->tty;
	struct screen		*s = &md->s;
	struct menu		*menu = md->menu;
	struct screen_write_ctx	 ctx;
	u_int			 i, px = md->px, py = md->py;

	screen_write_start(&ctx, s);
	screen_write_clearscreen(&ctx, 8);

	if (md->border_lines != BOX_LINES_NONE) {
		screen_write_box(&ctx, menu->width + 4, menu->count + 2,
		    md->border_lines, &md->border_style, menu->title);
	}

	screen_write_menu(&ctx, menu, md->choice, md->border_lines,
	    &md->style, &md->border_style, &md->selected_style);
	screen_write_stop(&ctx);

	for (i = 0; i < screen_size_y(&md->s); i++) {
		tty_draw_line(tty, s, 0, i, menu->width + 4, px, py + i,
		    &grid_default_cell, NULL);
	}
}

void
menu_free_cb(__unused struct client *c, void *data)
{
	struct menu_data	*md = data;

	if (md->item != NULL)
		cmdq_continue(md->item);

	if (md->cb != NULL)
		md->cb(md->menu, UINT_MAX, KEYC_NONE, md->data);

	screen_free(&md->s);
	menu_free(md->menu);
	free(md);
}

int
menu_key_cb(struct client *c, void *data, struct key_event *event)
{
	struct menu_data		*md = data;
	struct menu			*menu = md->menu;
	struct mouse_event		*m = &event->m;
	u_int				 i;
	int				 count = menu->count, old = md->choice;
	const char			*name = NULL;
	const struct menu_item		*item;
	struct cmdq_state		*state;
	enum cmd_parse_status		 status;
	char				*error;

	if (KEYC_IS_MOUSE(event->key)) {
		if (md->flags & MENU_NOMOUSE) {
			if (MOUSE_BUTTONS(m->b) != MOUSE_BUTTON_1)
				return (1);
			return (0);
		}
		if (m->x < md->px ||
		    m->x > md->px + 4 + menu->width ||
		    m->y < md->py + 1 ||
		    m->y > md->py + 1 + count - 1) {
			if (~md->flags & MENU_STAYOPEN) {
				if (MOUSE_RELEASE(m->b))
					return (1);
			} else {
				if (!MOUSE_RELEASE(m->b) &&
				    !MOUSE_WHEEL(m->b) &&
				    !MOUSE_DRAG(m->b))
					return (1);
			}
			if (md->choice != -1) {
				md->choice = -1;
				c->flags |= CLIENT_REDRAWOVERLAY;
			}
			return (0);
		}
		if (~md->flags & MENU_STAYOPEN) {
			if (MOUSE_RELEASE(m->b))
				goto chosen;
		} else {
			if (!MOUSE_WHEEL(m->b) && !MOUSE_DRAG(m->b))
				goto chosen;
		}
		md->choice = m->y - (md->py + 1);
		if (md->choice != old)
			c->flags |= CLIENT_REDRAWOVERLAY;
		return (0);
	}
	for (i = 0; i < (u_int)count; i++) {
		name = menu->items[i].name;
		if (name == NULL || *name == '-')
			continue;
		if (event->key == menu->items[i].key) {
			md->choice = i;
			goto chosen;
		}
	}
	switch (event->key & ~KEYC_MASK_FLAGS) {
	case KEYC_UP:
	case 'k':
		if (old == -1)
			old = 0;
		do {
			if (md->choice == -1 || md->choice == 0)
				md->choice = count - 1;
			else
				md->choice--;
			name = menu->items[md->choice].name;
		} while ((name == NULL || *name == '-') && md->choice != old);
		c->flags |= CLIENT_REDRAWOVERLAY;
		return (0);
	case KEYC_BSPACE:
		if (~md->flags & MENU_TAB)
			break;
		return (1);
	case '\011': /* Tab */
		if (~md->flags & MENU_TAB)
			break;
		if (md->choice == count - 1)
			return (1);
		/* FALLTHROUGH */
	case KEYC_DOWN:
	case 'j':
		if (old == -1)
			old = 0;
		do {
			if (md->choice == -1 || md->choice == count - 1)
				md->choice = 0;
			else
				md->choice++;
			name = menu->items[md->choice].name;
		} while ((name == NULL || *name == '-') && md->choice != old);
		c->flags |= CLIENT_REDRAWOVERLAY;
		return (0);
	case KEYC_PPAGE:
	case 'b'|KEYC_CTRL:
		if (md->choice < 6)
			md->choice = 0;
		else {
			i = 5;
			while (i > 0) {
				md->choice--;
				name = menu->items[md->choice].name;
				if (md->choice != 0 &&
				    (name != NULL && *name != '-'))
					i--;
				else if (md->choice == 0)
					break;
			}
		}
		c->flags |= CLIENT_REDRAWOVERLAY;
		break;
	case KEYC_NPAGE:
		if (md->choice > count - 6) {
			md->choice = count - 1;
			name = menu->items[md->choice].name;
		} else {
			i = 5;
			while (i > 0) {
				md->choice++;
				name = menu->items[md->choice].name;
				if (md->choice != count - 1 &&
				    (name != NULL && *name != '-'))
					i++;
				else if (md->choice == count - 1)
					break;
			}
		}
		while (name == NULL || *name == '-') {
			md->choice--;
			name = menu->items[md->choice].name;
		}
		c->flags |= CLIENT_REDRAWOVERLAY;
		break;
	case 'g':
	case KEYC_HOME:
		md->choice = 0;
		name = menu->items[md->choice].name;
		while (name == NULL || *name == '-') {
			md->choice++;
			name = menu->items[md->choice].name;
		}
		c->flags |= CLIENT_REDRAWOVERLAY;
		break;
	case 'G':
	case KEYC_END:
		md->choice = count - 1;
		name = menu->items[md->choice].name;
		while (name == NULL || *name == '-') {
			md->choice--;
			name = menu->items[md->choice].name;
		}
		c->flags |= CLIENT_REDRAWOVERLAY;
		break;
	case 'f'|KEYC_CTRL:
		break;
	case '\r':
		goto chosen;
	case '\033': /* Escape */
	case 'c'|KEYC_CTRL:
	case 'g'|KEYC_CTRL:
	case 'q':
		return (1);
	}
	return (0);

chosen:
	if (md->choice == -1)
		return (1);
	item = &menu->items[md->choice];
	if (item->name == NULL || *item->name == '-') {
		if (md->flags & MENU_STAYOPEN)
			return (0);
		return (1);
	}
	if (md->cb != NULL) {
	    md->cb(md->menu, md->choice, item->key, md->data);
	    md->cb = NULL;
	    return (1);
	}

	if (md->item != NULL)
		event = cmdq_get_event(md->item);
	else
		event = NULL;
	state = cmdq_new_state(&md->fs, event, 0);

	status = cmd_parse_and_append(item->command, NULL, c, state, &error);
	if (status == CMD_PARSE_ERROR) {
		cmdq_append(c, cmdq_get_error(error));
		free(error);
	}
	cmdq_free_state(state);

	return (1);
}

static void
menu_set_style(struct client *c, struct grid_cell *gc, const char *style,
    const char *option)
{
	struct style	 sytmp;
	struct options	*o = c->session->curw->window->options;

	memcpy(gc, &grid_default_cell, sizeof *gc);
	style_apply(gc, o, option, NULL);
	if (style != NULL) {
		style_set(&sytmp, &grid_default_cell);
		if (style_parse(&sytmp, gc, style) == 0) {
			gc->fg = sytmp.gc.fg;
			gc->bg = sytmp.gc.bg;
		}
	}
}

struct menu_data *
menu_prepare(struct menu *menu, int flags, int starting_choice,
    struct cmdq_item *item, u_int px, u_int py, struct client *c,
    enum box_lines lines, const char *style, const char *selected_style,
    const char *border_style, struct cmd_find_state *fs, menu_choice_cb cb,
    void *data)
{
	struct menu_data	*md;
	int			 choice;
	const char		*name;
	struct options		*o = c->session->curw->window->options;

	if (c->tty.sx < menu->width + 4 || c->tty.sy < menu->count + 2)
		return (NULL);
	if (px + menu->width + 4 > c->tty.sx)
		px = c->tty.sx - menu->width - 4;
	if (py + menu->count + 2 > c->tty.sy)
		py = c->tty.sy - menu->count - 2;

	if (lines == BOX_LINES_DEFAULT)
		lines = options_get_number(o, "menu-border-lines");

	md = xcalloc(1, sizeof *md);
	md->item = item;
	md->flags = flags;
	md->border_lines = lines;

	menu_set_style(c, &md->style, style, "menu-style");
	menu_set_style(c, &md->selected_style, selected_style,
	    "menu-selected-style");
	menu_set_style(c, &md->border_style, border_style, "menu-border-style");

	if (fs != NULL)
		cmd_find_copy_state(&md->fs, fs);
	screen_init(&md->s, menu->width + 4, menu->count + 2, 0);
	if (~md->flags & MENU_NOMOUSE)
		md->s.mode |= (MODE_MOUSE_ALL|MODE_MOUSE_BUTTON);
	md->s.mode &= ~MODE_CURSOR;

	md->px = px;
	md->py = py;

	md->menu = menu;
	md->choice = -1;

	if (md->flags & MENU_NOMOUSE) {
		if (starting_choice >= (int)menu->count) {
			starting_choice = menu->count - 1;
			choice = starting_choice + 1;
			for (;;) {
				name = menu->items[choice - 1].name;
				if (name != NULL && *name != '-') {
					md->choice = choice - 1;
					break;
				}
				if (--choice == 0)
					choice = menu->count;
				if (choice == starting_choice + 1)
					break;
			}
		} else if (starting_choice >= 0) {
			choice = starting_choice;
			for (;;) {
				name = menu->items[choice].name;
				if (name != NULL && *name != '-') {
					md->choice = choice;
					break;
				}
				if (++choice == (int)menu->count)
					choice = 0;
				if (choice == starting_choice)
					break;
			}
		}
	}

	md->cb = cb;
	md->data = data;
	return (md);
}

int
menu_display(struct menu *menu, int flags, int starting_choice,
    struct cmdq_item *item, u_int px, u_int py, struct client *c,
    enum box_lines lines, const char *style, const char *selected_style,
    const char *border_style, struct cmd_find_state *fs, menu_choice_cb cb,
    void *data)
{
	struct menu_data	*md;

	md = menu_prepare(menu, flags, starting_choice, item, px, py, c, lines,
	    style, selected_style, border_style, fs, cb, data);
	if (md == NULL)
		return (-1);
	server_client_set_overlay(c, 0, NULL, menu_mode_cb, menu_draw_cb,
	    menu_key_cb, menu_free_cb, NULL, md);
	return (0);
}
